exporter/sumologicexporter/sender.go

```diff
@@ -0,0 +1,277 @@
+	default:
+		return errors.New("unexpected pipeline")
+	}
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
+		return fmt.Errorf("error during sending data: %s", resp.Status)
+	}
+	return nil
+}
+
+// sendLogs sends logs in right format basing on the s.config.LogFormat
+func (s *sender) sendLogs(flds fields) ([]pdata.LogRecord, error) {
+	switch s.config.LogFormat {
+	case TextFormat:
+		return s.sendLogsTextFormat(flds)
+	case JSONFormat:
+		return s.sendLogsJSONFormat(flds)
+	default:
+		return nil, errors.New("unexpected log format")
+	}
+}
+
+// sendLogsTextFormat sends log records from the buffer as text data
+// and as the result of execution returns array of records which has not been
+// sent correctly and error
+func (s *sender) sendLogsTextFormat(flds fields) ([]pdata.LogRecord, error) {
+	var (
+		body strings.Builder
+		errs []error
+		// droppedRecords tracks all dropped records
+		droppedRecords []pdata.LogRecord
+		// currentRecords tracks records which are being actually processed
+		currentRecords []pdata.LogRecord
+	)
+
+	for _, record := range s.buffer {
+		sent, appended, err := s.appendAndSend(record.Body().StringVal(), LogsPipeline, &body, flds)
+		if err != nil {
+			errs = append(errs, err)
+			if sent {
+				droppedRecords = append(droppedRecords, currentRecords...)
+			}
+
+			if !appended {
+				droppedRecords = append(droppedRecords, record)
+			}
+		}
+
+		// If data was sent, cleanup the currentTimeSeries counter
+		if sent {
+			currentRecords = currentRecords[:0]
+		}
+
+		// If log has been appended to body, increment the currentTimeSeries
+		if appended {
+			currentRecords = append(currentRecords, record)
+		}
+	}
+
+	if err := s.send(LogsPipeline, strings.NewReader(body.String()), flds); err != nil {
+		errs = append(errs, err)
+		droppedRecords = append(droppedRecords, currentRecords...)
+	}
+
+	if len(errs) > 0 {
+		return droppedRecords, componenterror.CombineErrors(errs)
+	}
+	return droppedRecords, nil
+}
+
+// sendLogsJSONFormat sends log records from the buffer as json data
+// and as the result of execution returns array of records which has not been
+// sent correctly and error
+func (s *sender) sendLogsJSONFormat(flds fields) ([]pdata.LogRecord, error) {
+	var (
+		body strings.Builder
+		errs []error
+		// droppedRecords tracks all dropped records
+		droppedRecords []pdata.LogRecord
+		// currentRecords tracks records which are being actually processed
+		currentRecords []pdata.LogRecord
+	)
+
+	for _, record := range s.buffer {
+		data := s.filter.filterOut(record.Attributes())
+		data[logKey] = record.Body().StringVal()
+
+		nextLine, err := json.Marshal(data)
+		if err != nil {
+			droppedRecords = append(droppedRecords, record)
+			errs = append(errs, err)
+			continue
+		}
+
+		sent, appended, err := s.appendAndSend(bytes.NewBuffer(nextLine).String(), LogsPipeline, &body, flds)
```

Comment:
From this point and on, this looks just like `sendLogsTextFormat`. Can you refactor it to extract the common code into its own function?

Reply:
yup, done. significant improvement 👍